src/beanmachine/ppl/inference/predictive.py

```diff
@@ -37,7 +38,8 @@ def simulate(  # noqa: C901
         posterior: Optional[MonteCarloSamples] = None,
         num_samples: Optional[int] = None,
         vectorized: Optional[bool] = False,
-    ) -> MonteCarloSamples:
+        return_inference_data: bool = False,
+    ) -> Union[MonteCarloSamples, az.InferenceData]:
```

Comment:
Will the type checkers (pyre, mypy) be able to figure this out, or are we stuck with a union until we type cast it?

Reply:
Is this not the correct type for this function?

Reply:
I am guessing that type checking tools for Python are not particular comfortable with structural types (like disjoint unions and products). Would it be helpful here to decide whether we should find a way to make az.InferenceData a kind of subtype/special case of MonteCarloSample?

Reply:
There is still a type cast from a union to an explicit type (`MonteCarloSamples` for now) which would need to be added to all callers of `simulate`; it might be easier to define a new method and punt on backwards compatibility.